src/Engine/MapController.java

```diff
@@ -586,8 +564,30 @@ public void animationEnded(GameEvent event, boolean animEventQueueIsEmpty)
       }
       else
       {
-        // The animation for the last action just completed. Back to normal input mode.
-        changeInputMode(InputMode.INPUT);
+        // The animation for the last action just completed. If an AI is in control,
+        // fetch the next action. Otherwise, return control to the player.
+        if( myGame.activeCO.isAI() )
+        {
+          GameAction aiAction = myGame.activeCO.getNextAIAction(myGame.gameMap);
+          boolean endAITurn = false;
+          if( aiAction != null )
+          {
+            if( !executeGameAction(aiAction) )
+            {
+              // If aiAction fails to execute, the AI's turn is over. We don't want
+              // to waste time getting more actions if it can't build them properly.
+              System.out.println("WARNING! AI Action " + aiAction.toString() + " Failed to execute!");
+              endAITurn = true;
+            }
+          }
+          else { endAITurn = true; } // The AI can return a null action to signal the end of its turn.
+          if( endAITurn) startNextTurn();
+        }
+        else
+        {
+          // Back to normal input mode.
+          changeInputMode(InputMode.INPUT);
+        }
```

Comment:
In retrospect, it's pretty funny that I never thought to just put the handling code here, rather than calling the other function.

Reply:
Yeah, it was kindof a "well, duh" moment for me too.